netutil/netutil.go

```diff
@@ -0,0 +1,53 @@
+package netutil
+
+import (
+	"io"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/libp2p/go-libp2p"
+	"github.com/libp2p/go-libp2p-core/peer"
+	"github.com/multiformats/go-multiaddr"
+	"golang.org/x/net/context"
+)
+
+func TryAcquireLatency(ctx context.Context, ai peer.AddrInfo) (multiaddr.Multiaddr, int64, error) {
+	cctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+	defer cancel()
+	h, err := libp2p.New(cctx)
+	if err != nil {
+		return nil, 0, err
+	}
+
+	h.Peerstore().AddAddrs(ai.ID, ai.Addrs, time.Hour)
+	conn, err := h.Network().DialPeer(cctx, ai.ID)
+	if err != nil {
+		return nil, 0, err
+	}
+	remote := conn.RemoteMultiaddr()
+	start := time.Now()
+	var end time.Time
+	stream, err := conn.NewStream()
+	stream.SetReadDeadline(time.Now().Add(3 * time.Second))
+
+	wg := sync.WaitGroup{}
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		p := []byte{0x00}
+		_, serr := stream.Read(p)
+		if serr == io.EOF || serr == io.ErrUnexpectedEOF || strings.Contains(serr.Error(), "connection reset by peer") {
+			end = time.Now()
+		} else {
+			err = serr
+		}
+	}()
+
+	// send a write of random bytes, which should make the other side unhappy.
+	stream.Write([]byte("-latencytest-\n\x00"))
+	wg.Wait()
+	defer conn.Close()
+
```

Comment:
```go
if err != nil {
   return nil, 0, err
}
```

Reply:
we may still get a useful remote node multiaddr by this point, which is why i'm tempted to still return it even if the error indicates the measured latency shouldn't be trusted